### PR TITLE
[NDB_BVL_Instrument] fix support for JSONData instruments

### PR DESCRIPTION
**Pull request by ridz1208**
## Brief summary of changes
Fix a couple instances where the Instrument class does not support JSONData instruments by either looking for a specific field in the not-yet-populated instrument data object or using the `$this->table` variable

Testing

1. choose a JSONData=true instrument (you can turn `false` into `true` in the `bmi.meta` file for an easy instruement)
2. set he `Data` column to `NULL` directly in the flag table in the database
3. try entering the "Date of Administration" and other required fields on the top page and save.

Without this PR the save will fail claiming the `Date_taken` field does not exist in the instrument. With this PR, issue should be solved

### CODE REVIEW

**Comment by ridz1208** [comment]
@driusan @kongtiaowang I'm not sure why this is still failing. On my VM everything seems functional

**Comment by ridz1208** [comment]
@kongtiaowang on line 129 of `NDB_BVL_Instrument_Test.php` the get datadisctionary returns an empty array. It needs to be changed to return DictionaryItems I'm just not sure how to make mock items

**Comment by ridz1208** [comment]
@kongtiaowang @driusan after 17000 tries, I think I got it right

**Comment by ridz1208** [comment]
@zaliqarosli would you be able to test this one ?

**Comment by ridz1208** [comment]
blocking https://github.com/aces/Loris/pull/7809

**Comment by zaliqarosli** [comment]
i'm getting a different error on both this PR branch and main. is there a patch im missing?

```
[Tue Nov 16 16:31:49.714552 2021] [php7:notice] [pid 27212] [client 10.127.130.109:52076] [DEBUG] Module handle function called, referer: http://zrosli-dev.loris.ca/instruments/bmi/?candID=300001&sessionID=1&commentID=300001MTL0011221465332127
[Tue Nov 16 16:31:49.765644 2021] [php7:notice] [pid 27212] [client 10.127.130.109:52076] [DEBUG] Page bmi not found, referer: http://zrosli-dev.loris.ca/instruments/bmi/?candID=300001&sessionID=1&commentID=300001MTL0011221465332127
[Tue Nov 16 16:31:49.765726 2021] [php7:notice] [pid 27212] [client 10.127.130.109:52076] [DEBUG] Loading bmi , referer: http://zrosli-dev.loris.ca/instruments/bmi/?candID=300001&sessionID=1&commentID=300001MTL0011221465332127
[Tue Nov 16 16:31:49.814912 2021] [php7:notice] [pid 27212] [client 10.127.130.109:52076] [DEBUG] NDB_Page middleware called, referer: http://zrosli-dev.loris.ca/instruments/bmi/?candID=300001&sessionID=1&commentID=300001MTL0011221465332127
```

Screen Shot 2021-11-16 at 4 34 13 PM

**Comment by ridz1208** [comment]
@zaliqarosli is the BMI instruemnt set up on your VM ? (if not you need to copy it from raisinbread/instruments into project instruments)

**Comment by zaliqarosli** [comment]
@ridz1208 yeah the instrument loads fine for me

Screen Shot 2021-11-16 at 4 40 21 PM

I just get the 500 error with no useful messages in my apache log when I try to 'Save' data

**Comment by zaliqarosli** [comment]
@ridz1208 yes

**Comment by ridz1208** [comment]
@zaliqarosli if only we could know what the error is lol

**Comment by zaliqarosli** [comment]
i get 500 again and only a bunch of select statements in my log

```
[Tue Nov 16 17:28:17.716416 2021] [php7:notice] [pid 28082] [client 10.127.138.104:55978] [DEBUG] SELECT cs.ID as ParentID, child.ID as ChildID,\n                        cs.AllowMultiple, child.Name\n                FROM ConfigSettings cs\n                    LEFT JOIN ConfigSettings child ON (child.Parent=cs.ID)\n                WHERE cs.ID=:101, referer: http://zrosli-dev.loris.ca/instruments/bmi/?commentID=300001MTL0011221465332127&sessionID=1&candID=300001
[Tue Nov 16 17:28:17.717588 2021] [php7:notice] [pid 28082] [client 10.127.138.104:55978] [DEBUG] SELECT Value FROM Config WHERE ConfigID=:101 LIMIT 2, referer: http://zrosli-dev.loris.ca/instruments/bmi/?commentID=300001MTL0011221465332127&sessionID=1&candID=300001
[Tue Nov 16 17:28:17.719192 2021] [php7:notice] [pid 28082] [client 10.127.138.104:55978] [DEBUG] SELECT cs.ID as ParentID, child.ID as ChildID,\n                        cs.AllowMultiple, child.Name\n                FROM ConfigSettings cs\n                    LEFT JOIN ConfigSettings child ON (child.Parent=cs.ID)\n                WHERE cs.ID=:102, referer: http://zrosli-dev.loris.ca/instruments/bmi/?commentID=300001MTL0011221465332127&sessionID=1&candID=300001
[Tue Nov 16 17:28:17.720336 2021] [php7:notice] [pid 28082] [client 10.127.138.104:55978] [DEBUG] SELECT Value FROM Config WHERE ConfigID=:102 LIMIT 2, referer: http://zrosli-dev.loris.ca/instruments/bmi/?commentID=300001MTL0011221465332127&sessionID=1&candID=300001
[Tue Nov 16 17:28:17.722055 2021] [php7:notice] [pid 28082] [client 10.127.138.104:55978] [DEBUG] SELECT cs.ID as ParentID, child.ID as ChildID,\n                        cs.AllowMultiple, child.Name\n                FROM ConfigSettings cs\n                    LEFT JOIN ConfigSettings child ON (child.Parent=cs.ID)\n                WHERE cs.ID=:103, referer: http://zrosli-dev.loris.ca/instruments/bmi/?commentID=300001MTL0011221465332127&sessionID=1&candID=300001
[Tue Nov 16 17:28:17.723227 2021] [php7:notice] [pid 28082] [client 10.127.138.104:55978] [DEBUG] SELECT Value FROM Config WHERE ConfigID=:103 LIMIT 2, referer: http://zrosli-dev.loris.ca/instruments/bmi/?commentID=300001MTL0011221465332127&sessionID=1&candID=300001
[Tue Nov 16 17:28:17.724786 2021] [php7:notice] [pid 28082] [client 10.127.138.104:55978] [DEBUG] SELECT cs.ID as ParentID, child.ID as ChildID,\n                        cs.AllowMultiple, child.Name\n                FROM ConfigSettings cs\n                    LEFT JOIN ConfigSettings child ON (child.Parent=cs.ID)\n                WHERE cs.ID=:106, referer: http://zrosli-dev.loris.ca/instruments/bmi/?commentID=300001MTL0011221465332127&sessionID=1&candID=300001
[Tue Nov 16 17:28:17.725929 2021] [php7:notice] [pid 28082] [client 10.127.138.104:55978] [DEBUG] SELECT Value FROM Config WHERE ConfigID=:106 LIMIT 2, referer: http://zrosli-dev.loris.ca/instruments/bmi/?commentID=300001MTL0011221465332127&sessionID=1&candID=300001
```

**Comment by ridz1208** [comment]
@driusan i dont know what happened to exceptions but as you can see... impossible to debug

**Comment by ridz1208** [comment]
@zaliqarosli please test/review again. if possible with both a linst and non linst instrument

File: php/libraries/NDB_BVL_Instrument_LINST.class.inc [hunk 1]
```diff
@@ -86,7 +86,7 @@ class NDB_BVL_Instrument_LINST extends \NDB_BVL_Instrument
             $this->dictionary,
             [
                 new DictionaryItem(
-                    'Date_taken',
+                    $this->testName.'_Date_taken',
```
Comment (ridz1208) [inline php/libraries/NDB_BVL_Instrument_LINST.class.inc:89]: @driusan all changes made to this file it to make it at par with the `LorisFormDictionaryImpl.class.inc` which prefixes all dictionary items with the instrument name

**Review by zaliqarosli — CHANGES_REQUESTED** [review]
works exactly as should with a linst file.

with a non-linst file, the data saves fine and scores are correctly updated and saved.
BUT for some reason, on clicking 'Save Data' on the Top page, the page refreshes and takes you to the 'General Observations' (3rd) page, with a weird combination of both the top page and general observations page rendered. (this does not happen on main branch)

Screen Shot 2021-11-22 at 1 19 20 PM

File: php/libraries/NDB_BVL_Instrument.class.inc [hunk 2]
```diff
@@ -1403,12 +1403,23 @@ abstract class NDB_BVL_Instrument extends NDB_Page
...
     {
-        $allValues = $this->getInstanceData();
-
-        if (!array_key_exists($field, $allValues)) {
-            throw new \OutOfBoundsException("Invalid field for instrument");
+        $allValues  = $this->getInstanceData();
+        $dictionary = self::factory(
+            $this->loris,
+            $this->testName
+        )->getDataDictionary();
```
Comment (ridz1208) [inline php/libraries/NDB_BVL_Instrument.class.inc:1410]: @driusan `getDataDictionary()` modifies the `$this->page` value altering the active page of the instrument causing the weird page change on save.

I fixed it here simply by getting the data dictionary from another instance of the same instrument... I don't like the fact that getDataDictionary modifies the instrument on which it is called however. To be discussed

**Review by zaliqarosli — APPROVED** [review]
works well now

**Comment by ridz1208** [comment]
@zaliqarosli this PR has changed a little, could you please re-test only PHP instruments?

**Review by zaliqarosli — APPROVED** [review]
still works!

**Comment by ridz1208** [comment]
@driusan PLEASE MERGE BEFORE IT BREAKS AGAIN

**Pull request merged**